<a href="https://colab.research.google.com/github/tqnhu2407/Introduction_to_NLP/blob/main/source.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Encoding Language into Numbers

## Getting Started with Tokenization

In [1]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.text import Tokenizer

In [2]:
from bs4 import BeautifulSoup
import string

In [3]:
from tensorflow.keras.preprocessing.sequence import pad_sequences

In [4]:
import json

In [5]:
sentences = [
             'Today is a sunny day',
             'Today is a rainy day',
             'Is it sunny today?'
]

In [6]:
tokenizer = Tokenizer(num_words = 100)
tokenizer.fit_on_texts(sentences)
word_index = tokenizer.word_index
print(word_index)

{'today': 1, 'is': 2, 'a': 3, 'sunny': 4, 'day': 5, 'rainy': 6, 'it': 7}


## Turning Sentences into Sequences

In [7]:
sequences = tokenizer.texts_to_sequences(sentences)
print(sequences)

[[1, 2, 3, 4, 5], [1, 2, 3, 6, 5], [2, 7, 4, 1]]


In [8]:
test_data = [
 'Today is a snowy day',
 'Will it be rainy tomorrow?'
]

There is some strange words (not appeared in the vocab `sentences`)

In [9]:
test_sequences = tokenizer.texts_to_sequences(test_data)
print(test_sequences)

[[1, 2, 3, 5], [7, 6]]


Lost a lot of information right?

### Using out-of-vocabulary tokens

In [10]:
tokenizer = Tokenizer(num_words = 100, oov_token="<OOV>")
tokenizer.fit_on_texts(sentences)
word_index = tokenizer.word_index

test_sequences = tokenizer.texts_to_sequences(test_data)
print(word_index)
print(test_sequences)

{'<OOV>': 1, 'today': 2, 'is': 3, 'a': 4, 'sunny': 5, 'day': 6, 'rainy': 7, 'it': 8}
[[2, 3, 4, 1, 6], [1, 8, 1, 7, 1]]


### Understanding padding

We need all the data to be in the **same shape**.

In [11]:
sentences = [
 'Today is a sunny day',
 'Today is a rainy day',
 'Is it sunny today?',
 'I really enjoyed walking in the snow today'
]

In [12]:
tokenizer = Tokenizer(num_words=100, oov_token="<OOV>")
tokenizer.fit_on_texts(sentences)
word_index = tokenizer.word_index

sequences = tokenizer.texts_to_sequences(sentences)
print(sequences)

[[2, 3, 4, 5, 6], [2, 3, 4, 7, 6], [3, 8, 5, 2], [9, 10, 11, 12, 13, 14, 15, 2]]


In [13]:
padded = pad_sequences(sequences)
print(padded)

[[ 0  0  0  2  3  4  5  6]
 [ 0  0  0  2  3  4  7  6]
 [ 0  0  0  0  3  8  5  2]
 [ 9 10 11 12 13 14 15  2]]


The above is **prepadding**.

In [14]:
padded = pad_sequences(sequences, padding='post')
print(padded)

[[ 2  3  4  5  6  0  0  0]
 [ 2  3  4  7  6  0  0  0]
 [ 3  8  5  2  0  0  0  0]
 [ 9 10 11 12 13 14 15  2]]


We had too much padding huh?

In [15]:
padded = pad_sequences(sequences, padding='post', maxlen=6)
print(padded)

[[ 2  3  4  5  6  0]
 [ 2  3  4  7  6  0]
 [ 3  8  5  2  0  0]
 [11 12 13 14 15  2]]


The longest sequence was 

[ 9 10 11 12 13 14 15  2]

But now we lost some tokens at the begining.

In [16]:
padded = pad_sequences(sequences, padding='post', maxlen=6, truncating='post')
print(padded)

[[ 2  3  4  5  6  0]
 [ 2  3  4  7  6  0]
 [ 3  8  5  2  0  0]
 [ 9 10 11 12 13 14]]


# Removing Stopwords and Cleaning Text

* Remove HTML tags using BeautifulSoup.
* Have a list of stopwords.
* Remove punctuation using `string` library.




# Working with Real Data Sources

## Getting Text from Tensorflow Datasets

In [17]:
import tensorflow_datasets as tfds

In [18]:
imdb_sentences = []
train_data = tfds.as_numpy(tfds.load('imdb_reviews', split='train'))
for item in train_data:
  imdb_sentences.append(str(item['text']))

Shuffling and writing examples to /root/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteTTR8HS/imdb_reviews-train.tfrecord


Shuffling and writing examples to /root/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteTTR8HS/imdb_reviews-test.tfrecord


Shuffling and writing examples to /root/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteTTR8HS/imdb_reviews-unsupervised.tfrecord


Dataset imdb_reviews downloaded and prepared to /root/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


In [19]:
print(imdb_sentences[0])

b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it."


In [20]:
print(len(imdb_sentences))

25000


In [21]:
tokenizer = Tokenizer(num_words=5000)
tokenizer.fit_on_texts(imdb_sentences)
sequences = tokenizer.texts_to_sequences(imdb_sentences)

In [22]:
print(tokenizer.word_index)

{'the': 1, 'and': 2, 'a': 3, 'of': 4, 'to': 5, 'is': 6, 'br': 7, 'in': 8, 'it': 9, 'i': 10, 'this': 11, 'that': 12, 'was': 13, 'as': 14, 'for': 15, 'with': 16, 'movie': 17, 'but': 18, 'film': 19, "'s": 20, 'on': 21, 'you': 22, 'not': 23, 'are': 24, 'his': 25, 'he': 26, 'have': 27, 'be': 28, 'one': 29, 'all': 30, 'at': 31, 'by': 32, 'they': 33, 'an': 34, 'who': 35, 'so': 36, 'from': 37, 'like': 38, 'her': 39, "'t": 40, 'or': 41, 'just': 42, 'there': 43, 'about': 44, 'out': 45, "'": 46, 'has': 47, 'if': 48, 'some': 49, 'what': 50, 'good': 51, 'more': 52, 'very': 53, 'when': 54, 'she': 55, 'up': 56, 'can': 57, 'b': 58, 'time': 59, 'no': 60, 'even': 61, 'my': 62, 'would': 63, 'which': 64, 'story': 65, 'only': 66, 'really': 67, 'see': 68, 'their': 69, 'had': 70, 'were': 71, 'me': 72, 'well': 73, 'we': 74, 'than': 75, 'much': 76, 'been': 77, 'get': 78, 'bad': 79, 'will': 80, 'people': 81, 'do': 82, 'also': 83, 'other': 84, 'into': 85, 'because': 86, 'great': 87, 'first': 88, 'him': 89, 'how'

**br**: HTML tag which inserts a single line break.

Update the code to use BeautifulSoup to remove the HTML tags, add string
translation to remove the punctuation, and remove stopwords from the given list

In [23]:
stopwords = ['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', 'her', 'hers', 'herself', 'it', 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'nor', 'only', 'own', 'same', 'so', 'than', 'too', 'very', 's', 't', 'can', 'will', 'just', 'don', 'should', 'now']

In [24]:
table = str.maketrans('', '', string.punctuation)

In [25]:
imdb_sentences = []
train_data = tfds.as_numpy(tfds.load('imdb_reviews', split="train"))

In [26]:
for item in train_data:
    sentence = str(item['text'].decode('UTF-8').lower())
    soup = BeautifulSoup(sentence)
    sentence = soup.get_text() # remove HTML tags
    sentence = sentence.replace(",", " , ") # to avoid "annoying-conclusion" or "him/her"
    sentence = sentence.replace(".", " . ")
    sentence = sentence.replace("-", " - ")
    sentence = sentence.replace("/", " / ")
    words = sentence.split()
    filtered_sentence = ""
    for word in words:
        word = word.translate(table) # remove punctuation
        if word not in stopwords:
            filtered_sentence += word + " "
    imdb_sentences.append(filtered_sentence)

In [27]:
tokenizer = Tokenizer(num_words=25000)
tokenizer.fit_on_texts(imdb_sentences)
sequences = tokenizer.texts_to_sequences(imdb_sentences)

In [28]:
print(tokenizer.word_index)

{'movie': 1, 'film': 2, 'not': 3, 'one': 4, 'like': 5, 'good': 6, 'no': 7, 'time': 8, 'even': 9, 'would': 10, 'story': 11, 'really': 12, 'see': 13, 'well': 14, 'much': 15, 'bad': 16, 'get': 17, 'also': 18, 'people': 19, 'great': 20, 'first': 21, 'dont': 22, 'made': 23, 'movies': 24, 'make': 25, 'way': 26, 'films': 27, 'could': 28, 'characters': 29, 'think': 30, 'watch': 31, 'two': 32, 'character': 33, 'many': 34, 'seen': 35, 'life': 36, 'plot': 37, 'never': 38, 'acting': 39, 'little': 40, 'love': 41, 'best': 42, 'show': 43, 'know': 44, 'ever': 45, 'better': 46, 'end': 47, 'still': 48, 'man': 49, 'say': 50, 'scene': 51, 'scenes': 52, 'go': 53, 'something': 54, 'back': 55, 'im': 56, 'real': 57, 'watching': 58, 'years': 59, 'doesnt': 60, 'though': 61, 'actors': 62, 'old': 63, 'thing': 64, 'didnt': 65, 'work': 66, 'new': 67, 'another': 68, 'nothing': 69, 'funny': 70, 'actually': 71, '10': 72, 'makes': 73, 'director': 74, 'look': 75, 'find': 76, 'going': 77, 'part': 78, 'every': 79, 'lot': 

### Using the IMDb subwords datasets

In [29]:
(train_data, test_data), info = tfds.load('imdb_reviews/subwords8k', split = (tfds.Split.TRAIN, tfds.Split.TEST), as_supervised=True, with_info=True)

Shuffling and writing examples to /root/tensorflow_datasets/imdb_reviews/subwords8k/1.0.0.incompleteXU4VH4/imdb_reviews-train.tfrecord


Shuffling and writing examples to /root/tensorflow_datasets/imdb_reviews/subwords8k/1.0.0.incompleteXU4VH4/imdb_reviews-test.tfrecord


Shuffling and writing examples to /root/tensorflow_datasets/imdb_reviews/subwords8k/1.0.0.incompleteXU4VH4/imdb_reviews-unsupervised.tfrecord


Dataset imdb_reviews downloaded and prepared to /root/tensorflow_datasets/imdb_reviews/subwords8k/1.0.0. Subsequent calls will reuse this data.


In [30]:
encoder = info.features['text'].encoder
print(f'Vocabulary size: {encoder.vocab_size}')

Vocabulary size: 8185


In [31]:
print(encoder.subwords)

['the_', ', ', '. ', 'a_', 'and_', 'of_', 'to_', 's_', 'is_', 'br', 'in_', 'I_', 'that_', 'this_', 'it_', ' /><', ' />', 'was_', 'The_', 'as_', 't_', 'with_', 'for_', '.<', 'on_', 'but_', 'movie_', ' (', 'are_', 'his_', 'have_', 'film_', 'not_', 'ing_', 'be_', 'ed_', 'you_', ' "', 'it', 'd_', 'an_', 'he_', 'by_', 'at_', 'one_', 'who_', 'y_', 'from_', 'e_', 'or_', 'all_', 'like_', 'they_', '" ', 'so_', 'just_', 'has_', ') ', 'her_', 'about_', 'out_', 'This_', 'some_', 'ly_', 'movie', 'film', 'very_', 'more_', 'It_', 'would_', 'what_', 'when_', 'which_', 'good_', 'if_', 'up_', 'only_', 'even_', 'their_', 'had_', 'really_', 'my_', 'can_', 'no_', 'were_', 'see_', 'she_', '? ', 'than_', '! ', 'there_', 'get_', 'been_', 'into_', ' - ', 'will_', 'much_', 'story_', 'because_', 'ing', 'time_', 'n_', 'we_', 'ed', 'me_', ': ', 'most_', 'other_', 'don', 'do_', 'm_', 'es_', 'how_', 'also_', 'make_', 'its_', 'could_', 'first_', 'any_', "' ", 'people_', 'great_', 've_', 'ly', 'er_', 'made_', 'r_', 'B

In [32]:
sample_string = 'Today is a sunny day'

encoded_string = encoder.encode(sample_string)
print(f'Encoded string is {encoded_string}')

Encoded string is [6427, 4869, 9, 4, 2365, 1361, 606]


token = 6427 -> 6426th item in the array.

5 words are encoded into 7 tokens.

In [33]:
for t in encoded_string:
    print(f'token={t}: {encoder.subwords[t-1]}')

token=6427: Tod
token=4869: ay_
token=9: is_
token=4: a_
token=2365: sun
token=1361: ny_
token=606: day


In [34]:
original_string = encoder.decode(encoded_string)
print(original_string)

Today is a sunny day


## Getting Text from CSV Files

Temporarily skipped

## Getting Text from JSON Files

### Reading JSON files

In [35]:
sentences = []
labels = []
urls = []

In [36]:
with open('Sarcasm_Headlines_Dataset.json', 'r') as f:
    for line in f:
        obj = json.loads(line)
        sentence = obj['headline'].lower()
        sentence = sentence.replace(",", " , ")
        sentence = sentence.replace(".", " . ")
        sentence = sentence.replace("-", " - ")
        sentence = sentence.replace("/", " / ")
        soup = BeautifulSoup(sentence)
        sentence = soup.get_text() # remove HTML tags
        words = sentence.split()
        filtered_sentence = ""
        for word in words:
            word = word.translate(table) # remove punctuation
            if word not in stopwords: # remove stop words
                filtered_sentence = filtered_sentence + word + " "
        sentences.append(filtered_sentence)
        urls.append(obj['article_link'])
        labels.append(obj['is_sarcastic'])

FileNotFoundError: ignored

In [ ]:
len(labels)

In [ ]:
training_size = 23000
training_sentences = sentences[0:training_size]
testing_sentences = sentences[training_size:]
training_labels = labels[0:training_size]
testing_labels = labels[training_size:]

In [ ]:
tokenizer = Tokenizer(num_words=10000, oov_token="<OOV>")
tokenizer.fit_on_texts(training_sentences)

word_index = tokenizer.word_index

training_sequences = tokenizer.texts_to_sequences(training_sentences)
training_padded = pad_sequences(training_sequences, maxlen=10, padding='post', truncating='post')

In [ ]:
print(word_index)

In [ ]:
tokenizer.fit_on_texts(testing_sentences)
testing_sequences = tokenizer.texts_to_sequences(testing_sentences)
testing_padded = pad_sequences(testing_sequences, maxlen=10, padding='post', truncating='post')